src/config/AgaviXmlConfigParser.class.php

```diff
@@ -494,21 +494,25 @@ public static function executeCompilation(AgaviXmlConfigDomDocument $document, $
 	 */
 	public static function xinclude(AgaviXmlConfigDomDocument $document)
 	{
-		// replace %lala% directives in XInclude href attributes
-		foreach($document->getElementsByTagNameNS(self::NAMESPACE_XINCLUDE_2001, 'include') as $element) {
+		$elements = $document->getElementsByTagNameNS(self::NAMESPACE_XINCLUDE_2001, 'include');
+		$length = $elements->length;
+		for($i = 0; $i < $length; $i++) {
+			$element = $elements->item($i);
 			if($element->hasAttribute('href')) {
 				$attribute = $element->getAttributeNode('href');
 				$parts = explode('#', $attribute->nodeValue, 2);
 				$parts[0] = str_replace('\\', '/', AgaviToolkit::expandDirectives($parts[0]));
-				$attribute->nodeValue = implode('#', $parts);
+				$attribute->nodeValue = rawurlencode($parts[0]) . (isset($parts[1]) ? '#' . $parts[1] : '');
 				if(strpos($parts[0], '*') !== false || strpos($parts[0], '{') !== false) {
 					$glob = glob($parts[0], GLOB_BRACE | GLOB_NOSORT);
 					if($glob) {
 						$glob = array_unique($glob); // it could be that someone used /path/to/{Foo,*}/burp.xml so Foo would come before all others, that's why we need to remove duplicates as the * would match Foo again
+						sort($glob);
 						foreach($glob as $path) {
 							$new = $element->cloneNode(true);
-							$new->setAttribute('href', $path . (isset($parts[1]) ? '#' . $parts[1] : ''));
+							$new->setAttribute('href', rawurlencode($path) . (isset($parts[1]) ? '#' . $parts[1] : ''));
 							$element->parentNode->insertBefore($new, $element);
+							++$i;
```

Comment:
True, I'm just not used to see comments in Agavi code :p.

It was actually buggy before, because the `DOMNodeList` returned by `getElementsByTagNameNS()` has some auto-updating magic (see http://php.net/manual/en/class.domnodelist.php#83178). So we need to use a `for` loop and the `++$i` to skip the newly added elements.